src/serviceTokenProvider.js

```diff
@@ -0,0 +1,71 @@
+const jwtManager = require('jsonwebtoken');
+
+class ServiceTokenProvider {
+  /**
+   * Constructor
+   * @param {Object} httpClient used to perform http calls
+   * @param {Object} kmsClient used to decrypt the encrypted secret
+   * @param {Object} configuration configuration object
+   * @param {String} configuration.clientId client ID
+   * @param {String} configuration.encryptedClientSecret base64 encoded encrypted secret
+   * @param {String} configuration.audience jwt audience
+   * @param {String} configuration.tokenEndpoint endpoint to resolve token from
+   */
+  constructor(httpClient, kmsClient, configuration = {}) {
+    this.httpClient = httpClient;
+    this.kmsClient = kmsClient;
+    if (!configuration.clientId) {
+      throw new Error('Configuration error: missing required property "clientId"');
+    }
+    if (!configuration.encryptedClientSecret) {
+      throw new Error('Configuration error: missing required property "encryptedClientSecret"');
+    }
+    if (!configuration.audience) {
+      throw new Error('Configuration error: missing required property "audience"');
+    }
+    if (!configuration.tokenEndpoint) {
+      throw new Error('Configuration error: missing required property "tokenEndpoint"');
+    }
+    this.configuration = configuration;
+    this.currentTokenPromise = null;
+  }
+
+  /**
+   * Get access token. Subsequent calls are cached and the token is renewed only if it is expired.
+   * @returns {Promise<String>} access token
+   */
+  async getToken() {
+    if (!this.currentTokenPromise) {
+      return (this.currentTokenPromise = this.getTokenWithoutCache());
+    }
+
+    try {
+      let credentials = await this.currentTokenPromise;
+      if (!credentials || jwtManager.decode(credentials).exp < Date.now() / 1000) {
```

Comment:
I think we'll have a race-condition here where the token isn't expired yet, but as the call is made, it expires. It's a super-rare case, but removing 1-10 seconds or so for this check might give us slightly less cornercases. Low priority, especially since we didn't run into that in our own code yet - but good to think of.

Reply:
Makes sense.